### PR TITLE
Provide dynamic defaults via validators

### PR DESCRIPTION
**Pull request by SmileyChris**
Rather than calculating the defaults for `BASE_DIR`, `WSGI_APPLICATION`, and `ROOT_URLCONF` in `SetUp.configure`, calculate them dynamically on `PydanticSettings`.

Also part of this change is the ability to have a settings instance referenced in the `DJANGO_SETTINGS_MODULE` rather than just a class.

### CODE REVIEW

**Comment by joshourisman** [comment]
This looks pretty awesome! I need to find some time to sit down and actually look at it, but hopefully I'll be able to merge it in the next few days.

**Comment by SmileyChris** [comment]
Also, these three settings should really have a default of None so you can instantiate a settings object without typing warnings. I've done that in one of the commits in #103 (5419b11efa49a6643674114f15fd1d07cfebae65).

**Comment by joshourisman** [comment]
Ok, yeah, this seems pretty great; my initial implementation was definitely a bit hackey. Not entirely sure why the tests aren't running in the workflow, but they're all passing when I run them manually...

File: pydantic_settings/settings.py [hunk 1]
```diff
@@ -345,3 +326,34 @@ def remove_empty_database_dicts(cls, v):
...
+                path = path.parent
+            for part in module.__name__.split("."):
+                path = path.parent
+            values["BASE_DIR"] = path
+
+        base_module = None
+        if module:
+            base_module = module.__name__.split(".", 1)[0]
+        elif base_dir:
+            base_module = base_dir.name
```
Comment (SmileyChris) [inline pydantic_settings/settings.py:353]: `base_module` should probably only be calculated from `base_dir` (which should also be set after `values["BASE_DIR"] = path`), eliminating the need of the `if module` tree and ensuring BASE_DIR is the source of truth if provided.

**Pull request closed without merging**